tests/test_templates_broadcast.py

```diff
@@ -0,0 +1,243 @@
+
+
+@template
+def ConstantTemplateDouble(wires):
+    T(wires=wires[0])
+    CNOT(wires=wires)
+
+
+@template
+def ParametrizedTemplateDouble(par1, par2, wires):
+    CRX(par1, wires=wires)
+    RY(par2, wires=wires[0])
+
+
+@template
+def KwargTemplateDouble(par, wires, a=True):
+    if a:
+        T(wires=wires[0])
+    CRX(par, wires=wires)
+
+
+TARGET_OUTPUTS = [("single", [pi, pi, pi / 2, 0], RX, [1, 1, 0, -1]),
+                  ("double", [pi / 2, pi / 2], CRX, [-1, 0, -1, 0]),
+                  ("double", None, CNOT, [-1, 1, -1, 1]),
+                  ("double_odd", [pi / 2], CRX, [-1, -1, 0, -1]),
+                  ]
+
+CV_TARGET_OUTPUTS = [("single", [[0.1, 0.0], [0.2, 0.0], [0.3, 0.0], [0.4, 0.0]], Displacement, [2.2, 2.4, 2.6, 2.8],
+                      qml.X),
+                     ("double", [[pi / 4, 0.0], [pi / 4, 0.0]], Beamsplitter, [0, 2, 0, 2],
+                      qml.NumberOperator),
+                     ]
+
+GATE_PARAMETERS = [("single", RX, [[0.1], [0.2], [0.3]]),
+                   ("single", Rot, [[0.1, 0.2, 0.3], [0.3, 0.2, 0.1], [0.3, 0.2, -0.1]]),
+                   ("single", T, [[], [], []]),
+                   ("double", CRX, [[0.1]]),
+                   ("double", CRot, [[0.1, 0.2, 0.3]]),
+                   ("double", CNOT, [[]]),
+                   ("double_odd", CRX, [[0.1]]),
+                   ("double_odd", CRot, [[0.3, 0.2, 0.1]]),
+                   ("double_odd", CNOT, [[]]),
+                   ]
+
+
+class TestConstructorBroadcast:
+    """Tests the broadcast template constructor."""
+
+    @pytest.mark.parametrize("structure, unitary, parameters", GATE_PARAMETERS)
+    def test_correct_queue_for_gate_unitary(self, structure, unitary, parameters):
+        """Tests that correct gate queue is created when 'block' is a single gate."""
+
+        with qml.utils.OperationRecorder() as rec:
+            broadcast(block=unitary, structure=structure, wires=range(3), parameters=parameters)
+
+        for gate in rec.queue:
+            assert isinstance(gate, unitary)
+
+    @pytest.mark.parametrize("structure, unitary, gates, parameters",
+                             [("single", ParametrizedTemplate, [RX, RY], [[0.1, 1], [0.2, 1], [0.1, 1]]),
+                              ("single", ConstantTemplate, [T, S], [[], [], []]),
+                              ("double", ParametrizedTemplateDouble, [CRX, RY], [[0.1, 1]]),
+                              ("double", ConstantTemplateDouble, [T, CNOT], [[]]),
+                              ])
+    def test_correct_queue_for_template_block(self, structure, unitary, gates, parameters):
+        """Tests that correct gate queue is created when 'block' is a template."""
+
+        with qml.utils.OperationRecorder() as rec:
+            broadcast(block=unitary, structure=structure, wires=range(3), parameters=parameters)
+
+        for idx, gate in enumerate(rec.queue):
+            i = idx % 2
+            assert isinstance(gate, gates[i])
```

Comment:
It took me some time to parse the logic here, just as a suggestion you could also go with a more verbose version:
```python
first_gate = gates[0]
second_gate = gates[1]
for idx, gate in enumerate(rec.queue):
    if idx % 2 == 0:
        assert isinstance(gate, first_gate)
    else:
        assert isinstance(gate, second_gate)
```

Reply:
Thanks, implemented the more readable version!